stablehlo/dialect/TypeInference.cpp

```diff
@@ -571,62 +571,64 @@ LogicalResult inferConcatenateOp(Optional<Location> location, ValueRange inputs,
     }
   }
 
-  auto firstType = inputs[0].getType().cast<ShapedType>();
-  auto outElement = firstType.getElementType();
-
-  // Find the first ranked input to determine the output rank.
-  for (auto type : inputs.getTypes()) {
-    auto shapedType = type.cast<ShapedType>();
-    if (shapedType.hasRank()) {
-      firstType = shapedType;
-      break;
-    }
-  }
-
-  // If all inputs are unranked, the result must be unranked.
-  if (!firstType.hasRank()) {
-    inferredReturnTypes.push_back(UnrankedTensorType::get(outElement));
+  auto elementType = inputs[0].getType().cast<ShapedType>().getElementType();
+  if (!firstRankedType) {
+    inferredReturnTypes.push_back(UnrankedTensorType::get(elementType));
     return success();
   }
 
-  auto outShape = llvm::to_vector<6>(firstType.getShape());
-
-  // Determine what the non-concatenate dimensions should be.
-  for (auto type : inputs.getTypes()) {
-    auto shapedTy = type.cast<ShapedType>();
-    if (!shapedTy.hasRank()) {
-      continue;
-    }
-
-    for (const auto& it : llvm::enumerate(shapedTy.getShape())) {
-      // If a dimension is not dynamic, the output shape should match.
-      if (ShapedType::isDynamic(outShape[it.index()])) {
-        outShape[it.index()] = it.value();
+  // Infer the most specific (size, bound) of all dimenstions of the return type
+  auto rank = firstRankedType.getRank();
+  SmallVector<int64_t> inferredSizes(rank, ShapedType::kDynamicSize);
+  SmallVector<int64_t> inferredBounds(rank, ShapedType::kDynamicSize);
+  // Note: for the concatenate dimension, 0 should be the identical element:
+  // Any dim size can keep unchanged when concatenated with 0
+  inferredSizes[dimension] = 0;
+  bool anyInputHaveBounds = false;
+
+  // Note: unranked input types can't be ignored, consider these input types:
+  // c0: (<5x?xf32>, <*xf32>) with concat dim 0 should infer <?x?xf32>
+  // c1: (<5x?xf32>, <*xf32>) with concat dim 1 should infer <5x?xf32>
+  // Instead, they should be replaced with dynamic tensors: tensor<?x...?x>
+  for (const auto& it : llvm::enumerate(inputs.getTypes())) {
+    RankedTensorType rankedType = it.value().dyn_cast<RankedTensorType>();
+    SmallVector<int64_t> bounds;
+    if (rankedType)
+      bounds = to_vector(encodingToBounds(rankedType.getEncoding()));
+    if (!bounds.empty()) anyInputHaveBounds = true;
+
+    // Infer each dim for current rankedType (from inputTypes[1:end])
```

Comment:
I don't think that "from inputTypes[1:end]" applies anymore.

Reply:
Done.